### PR TITLE
Bug 1863319 - Fenix: fix flaky tests in TabbedBrowsingTest

### PR DESCRIPTION
**Pull request by sv-ohorvath**
Split the large tests into smaller ones. 
Matched them with new manual tests in TestRail.
Added more logging for easier debugging.
Removed keyboard visibility verification.
All of which improved the crash rates, from ~10-15/100 runs to 2-3/100 runs.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1863319
https://bugzilla.mozilla.org/show_bug.cgi?id=1853625
https://bugzilla.mozilla.org/show_bug.cgi?id=1857875

### CODE REVIEW

File: fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/ext/WaitNotNull.kt [hunk 1]
```diff
@@ -17,4 +18,8 @@ import org.mozilla.fenix.helpers.TestAssetHelper
 fun UiDevice.waitNotNull(
     searchCondition: SearchCondition<*>,
     waitTime: Long = TestAssetHelper.waitingTime,
-) = assertNotNull(wait(searchCondition, waitTime))
+) {
+    Log.d("TestLog", "Wait not null: $searchCondition")
```
Comment (jjSDET) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/ext/WaitNotNull.kt:22]: Should modify your log tag to match the rest of the logging statements and change it to `mozTestLog` here and the logging statement below

**Review by jjSDET — CHANGES_REQUESTED** [review]
While not necessary to change here, consider whether our ad-hoc logging should be consistent with `Log.i` or `Log.d`. 

While, functionally,  there's little difference, we can more easily filter on a single `log level` for custom logs later. Not an issue either way, afaik. I'm not sure what the trade-offs, if any, exist if we choose `information` or `debug` level.

File: fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/ext/WaitNotNull.kt (hunk 1 above)
Comment (sv-ohorvath) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/ext/WaitNotNull.kt:22]: Yes, I missed that detail. Thanks for catching it.

**Review by AndiAJ — APPROVED** [review]

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/TabbedBrowsingTest.kt [hunk 2]
```diff
@@ -54,10 +54,10 @@ class TabbedBrowsingTest {
 
     @get:Rule
     val activityTestRule = HomeActivityIntentTestRule.withDefaultSettingsOverrides(skipOnboarding = true)
-
-    @Rule
-    @JvmField
-    val retryTestRule = RetryTestRule(3)
+//
+//    @Rule
```
Comment (AndiAJ) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/TabbedBrowsingTest.kt:58]: Don't forget to restore the retryRule

File: fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/MatcherHelper.kt [hunk 3]
```diff
@@ -17,17 +18,25 @@ import org.mozilla.fenix.helpers.TestHelper.mDevice
...
+        Log.i("MozTestLog", "Looking for item with text: $itemText")
+        return mDevice.findObject(UiSelector().text(itemText))
+    }
 
-    fun itemWithDescription(description: String) =
-        mDevice.findObject(UiSelector().descriptionContains(description))
+    fun itemWithDescription(description: String): UiObject {
+        Log.i("MozTestLog", "Looking for item with description: $description")
+        return mDevice.findObject(UiSelector().descriptionContains(description))
+    }
```
Comment (AaronMT) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/MatcherHelper.kt:39]: Same comment here as Andi's PR
https://github.com/mozilla-mobile/firefox-android/pull/4390#discussion_r1386762233

File: fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/ext/WaitNotNull.kt [hunk 4]
```diff
@@ -17,4 +18,8 @@ import org.mozilla.fenix.helpers.TestAssetHelper
 fun UiDevice.waitNotNull(
     searchCondition: SearchCondition<*>,
     waitTime: Long = TestAssetHelper.waitingTime,
-) = assertNotNull(wait(searchCondition, waitTime))
+) {
+    Log.i("MozTestLog", "Wait not null: $searchCondition")
+    assertNotNull(wait(searchCondition, waitTime))
+    Log.i("MozTestLog", "Found $searchCondition not null")
+}
```
Comment (AaronMT) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/helpers/ext/WaitNotNull.kt:25]: See above comment, I'd just import the tag here for these

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt [hunk 5]
```diff
@@ -347,6 +347,7 @@ class NavigationToolbarRobot {
                         ViewActions.click(),
                     ),
                 )
+            Log.i("MozTestLog", "Clicked the tab shortcut New private tab button.")
```
Comment (AaronMT) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:350]: Same here

**Review by AaronMT — CHANGES_REQUESTED** [review]

**Comment by DonalMe** [comment]
@sv-ohorvath just a heads up that Mergify won't automatically merge this in the current state.
It needs `#changes-requested-reviews-by=0`, but there are 2 pending reviewers listed under "changes required".

**Pull request merged**